File: web/bat_easing.js
/**
 * BAT — keyframe easing shared by the animated editors' live previews.
 *
 * Mirror of bat_easing.py: a keyframe's `ease` names the curve used on the way
 * OUT of it, towards the next key; a key without one is linear, so existing
 * workflows interpolate exactly as before. The two tables must stay identical
 * or the preview and the render disagree mid-segment (tests/verify_easing.py).
 */

export const EASES = ["linear", "ease_in", "ease_out", "ease_in_out", "hold"];
export const DEFAULT_EASE = "linear";

/** Labels for the editors' ease pickers, in EASES order. */
export const EASE_LABELS = {
    linear: "Linear",
    ease_in: "Ease in",
    ease_out: "Ease out",
    ease_in_out: "Ease in-out",
    hold: "Hold",
};

/** The ease of a keyframe object (or the default for anything else). */
export function easeName(key) {
    const name = key && typeof key === "object" ? key.ease : undefined;
    return EASES.includes(name) ? name : DEFAULT_EASE;
}

/** Map a segment fraction t in [0, 1] through the named curve. */
export function applyEase(t, name) {
    t = t <= 0 ? 0 : t >= 1 ? 1 : t;
    switch (name) {
        case "ease_in":     return t * t;
        case "ease_out":    return 1 - (1 - t) * (1 - t);
        case "ease_in_out": return t * t * (3 - 2 * t);
        case "hold":        return 0;
        default:            return t;
    }
}

File: bat_easing.py
"""
BAT — keyframe easing shared by the animated editors.

Animated Crop, Animated Grade and Roto interpolate between keyframes. Until now
that was linear only. Each keyframe may now carry an ``ease`` naming the curve
used on the way OUT of it, towards the next key (After Effects' convention: the
outgoing segment belongs to the earlier key). A key without one is ``linear``,
so every saved workflow interpolates exactly as before.

``web/bat_easing.js`` is the same table for the editors' live previews; the two
must stay identical, or the preview and the render disagree mid-segment.
tests/verify_easing.py pins that.
"""

EASES = ("linear", "ease_in", "ease_out", "ease_in_out", "hold")
DEFAULT_EASE = "linear"


def ease_name(key):
    """The ease of a keyframe dict (or the default for anything else)."""
    name = key.get("ease") if isinstance(key, dict) else None
    return name if name in EASES else DEFAULT_EASE


def apply_ease(t, name):
    """Map a segment fraction t in [0, 1] through the named curve.

    ``hold`` keeps the earlier key's value until the next key (a step);
    the in/out curves are quadratic, ``ease_in_out`` is smoothstep.
    """
    t = 0.0 if t <= 0.0 else 1.0 if t >= 1.0 else float(t)
    if name == "ease_in":
        return t * t
    if name == "ease_out":
        return 1.0 - (1.0 - t) * (1.0 - t)
    if name == "ease_in_out":
        return t * t * (3.0 - 2.0 * t)
    if name == "hold":
        return 0.0
    return t
